### PR TITLE
API-11433: Handle Okta interactions on consumer signup

### PR DESCRIPTION
**Pull request by SlakrHakr**
https://vajira.max.gov/browse/API-11433

Handle Okta interactions that happen today in developer-portal-backend when a consumer submits the sandbox signup form in the developer-portal.

### CODE REVIEW

File: spec/support/vcr_cassettes/okta/consumer_signup_200.yml [hunk 1]
```diff
@@ -0,0 +1,326 @@
+---
+http_interactions:
+- request:
+    method: post
+    uri: https://dev-87720571-admin.okta.com/api/v1/apps
+    body:
+      encoding: UTF-8
+      string: '{"name":"oidc_client","label":"LPBDougFunnie-1639514168","signOnMode":"OPENID_CONNECT","credentials":{"oauthClient":{"autoKeyRotation":true,"token_endpoint_auth_method":"client_secret_basic"}},"visibility":{"autoSubmitToolbar":false,"hide":{"iOS":true,"web":true},"appLinks":{"login":false}},"settings":{"oauthClient":{"application_type":"web","consent_method":"REQUIRED","grant_types":["authorization_code","refresh_token"],"initiate_login_uri":"https://sandbox-api.va.gov/oauth2/redirect/","redirect_uris":["http://localhost:3000/callback","https://sandbox-api.va.gov/oauth2/redirect/"],"response_types":["code"]}}}'
```
Comment (SlakrHakr) [inline spec/support/vcr_cassettes/okta/consumer_signup_200.yml:8]: I figured this sort of request wouldnt be that big of a deal since its my own personal developer okta account that can be discarded at any point.

File: app/services/okta_service.rb [hunk 2]
```diff
@@ -11,4 +11,72 @@ def list_applications
...
+      @client.update_authorization_server_policy(auth_server_id, default_policy.id, default_policy)
+    end
+
+    application
+  end
+
+  private
+
+  def consumer_name(user)
+    "[LPB]#{user.consumer.organization}#{user.last_name}".gsub(/\W/, '')
```
Comment (FonzMP) [inline app/services/okta_service.rb:46]: do we need the `[]`,  the .gsub removes these. Rubocop fixed mine, as it needs a separate wrapper for the gsub to execute on... Also, should we match them for consistency or is there any concern, since it's temporary?
For reference: `"LPB-#{"#{organization}#{last_name}".gsub(/\W/, '')}"`

File: spec/support/vcr_cassettes/okta/consumer_signup_200.yml (hunk 1 above)
Comment (FonzMP) [reply spec/support/vcr_cassettes/okta/consumer_signup_200.yml:8]: I think I'm fine with it. I don't believe there is anything in there that's risky to commit. 

My only thought on this is, would it be worth it to spin up a dedicated Okta instance for LPB development so we share it across instead of having everyone's Okta Dev instance blasted throughout the repo? Plus it would just be one to remove/delete if something does happen?

**Comment by FonzMP** [comment]
Just the few comments above and we need to get the deployment repo squared away before merging.. other than that I think I'm good with it.

**Comment by FonzMP** [comment]
Oh, are you seeing the Faraday warning in rspec as well? Wonder if we can get those removed... 
```
...................WARNING: `Faraday::Connection#authorization` is deprecated; it will be removed in version 2.0.
While initializing your connection, use `#request(:authorization, ...)` instead.
See https://lostisland.github.io/faraday/middleware/authentication for more usage info.
...............................WARNING: `Faraday::Connection#authorization` is deprecated; it will be removed in version 2.0.
While initializing your connection, use `#request(:authorization, ...)` instead.
See https://lostisland.github.io/faraday/middleware/authentication for more usage info.
..............................................

Finished in 3.1 seconds (files took 4.78 seconds to load)
96 examples, 0 failures
```

File: spec/support/vcr_cassettes/okta/consumer_signup_200.yml (hunk 1 above)
Comment (SlakrHakr) [reply spec/support/vcr_cassettes/okta/consumer_signup_200.yml:8]: I'd vote we cross that bridge when we come to it, probably when/if we need to add additional okta functionality to this tool.

**Comment by SlakrHakr** [comment]
> Oh, are you seeing the Faraday warning in rspec as well? Wonder if we can get those removed...
> 
> ```
> ...................WARNING: `Faraday::Connection#authorization` is deprecated; it will be removed in version 2.0.
> While initializing your connection, use `#request(:authorization, ...)` instead.
> See https://lostisland.github.io/faraday/middleware/authentication for more usage info.
> ...............................WARNING: `Faraday::Connection#authorization` is deprecated; it will be removed in version 2.0.
> While initializing your connection, use `#request(:authorization, ...)` instead.
> See https://lostisland.github.io/faraday/middleware/authentication for more usage info.
> ..............................................
> 
> Finished in 3.1 seconds (files took 4.78 seconds to load)
> 96 examples, 0 failures
> ```

We discussed this offline and discovered it was because idme_group_id was missing from Alex's local configuration.
Documentation has been added to the README file to hopefully help get okta setup properly locally:
https://github.com/department-of-veterans-affairs/lighthouse-platform-backend/pull/43/commits/f97d374ca8a0b82c9aa4bd1b9d6f581700315bd8

File: app/services/okta_service.rb (hunk 2 above)
Comment (SlakrHakr) [reply app/services/okta_service.rb]: https://github.com/department-of-veterans-affairs/lighthouse-platform-backend/pull/43/commits/bdc2cf7250d6ac2389f7e70545251307b9eb0df4

**Comment by FonzMP** [comment]
Renaming the Okta Username with `LPB-` is all good, one weird thing I noticed was trying to send the same request gave me a generalized Okta error. It's functioning with the same process as the current dev portal and we don't see an issue there. Curious if you're seeing the same thing? 

When I tried to submit the same request via Postman, it was erroring out with a [generalized Okta error](https://developer.okta.com/docs/reference/error-codes/#E0000001). It wasn't until I deleted the applications in Okta it worked. Weird. Glad to sync on this offline.

**Comment by SlakrHakr** [comment]
> When I tried to submit the same request via Postman, it was erroring out with a generalized Okta error. It wasn't until I deleted the applications in Okta it worked.

Yeah, heres the full "generalSettings" error (that im seeing sometimes as well)
```
{:errorCode=>"E0000001",
 :errorSummary=>"Api validation failed: generalSettings",
 :errorLink=>"E0000001",
 :errorId=>"oaeNvt_bX_iRJmalpOLhCCIYA",
 :errorCauses=>
  [{:errorSummary=>
     "Cannot create application instance. Maximum number of instances has been reached."}]}
```
and yeah, deleting the applications fixes it just fine for me.

TLDR: This is most likely just due to using developer accounts locally, as far as I can tell.

**Comment by FonzMP** [comment]
functionally I'm OK with this now. Only catch is I think we update the `apis_list.csv` to include the needed values.

```
api_name,environment,metadata_url,api_ref,acl_ref,auth_server_access_key
fancy-service,dev,https://dev-api.va.gov/services/fancy-service/docs/v0/api,fancy,phan-sea
ohauth-service,dev,https://dev-api.va.gov/services/ohauth-possibly/docs/v0/api,ohAuth,,AUTHZ_SERVER_DEFAULT
```

**Review by FonzMP — APPROVED** [review]
:shipit:
